File: sdk/python/lifeos_sdk.py
"""
LifeOS Python SDK
Allows third-party developers to build plugins and agents
Version: 1.0.0
"""

import requests
import hashlib
import hmac
import secrets
from typing import Dict, List, Optional, Any
from datetime import datetime
from urllib.parse import urlencode


class LifeOSSDK:
    """LifeOS SDK for building plugins and agents"""
    
    def __init__(self, config: Dict[str, str]):
        """
        Initialize the SDK
        
        Args:
            config: Configuration dictionary with:
                - api_url: LifeOS API URL
                - client_id: OAuth client ID
                - client_secret: OAuth client secret
                - redirect_uri: OAuth redirect URI
        """
        self.api_url = config.get('api_url', 'http://localhost:8000')
        self.client_id = config['client_id']
        self.client_secret = config['client_secret']
        self.redirect_uri = config['redirect_uri']
        self.access_token: Optional[str] = None
        self.refresh_token: Optional[str] = None
        
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'User-Agent': 'LifeOS-SDK-Python/1.0.0'
        })
    
    # =========================================================================
    # AUTHENTICATION
    # =========================================================================
    
    def get_authorization_url(self, state: Optional[str] = None) -> str:
        """
        Get OAuth authorization URL
        
        Args:
            state: Random state for CSRF protection
            
        Returns:
            Authorization URL
        """
        if not state:
            state = secrets.token_hex(16)
        
        params = {
            'response_type': 'code',
            'client_id': self.client_id,
            'redirect_uri': self.redirect_uri,
            'scope': 'read:nodes write:nodes register:agent send:events',
            'state': state,
        }
        
        return f"{self.api_url}/oauth/authorize?{urlencode(params)}"
    
    def get_access_token(self, code: str) -> Dict[str, Any]:
        """
        Exchange authorization code for access token
        
        Args:
            code: Authorization code
            
        Returns:
            Token response dictionary
        """
        try:
            response = requests.post(
                f"{self.api_url}/oauth/token",
                json={
                    'grant_type': 'authorization_code',
                    'code': code,
                    'client_id': self.client_id,
                    'client_secret': self.client_secret,
                    'redirect_uri': self.redirect_uri,
                }
            )
            response.raise_for_status()
            
            data = response.json()
            self.access_token = data['access_token']
            self.refresh_token = data.get('refresh_token')
            
            self._update_session_auth()
            
            return data
        except requests.exceptions.RequestException as e:
            raise Exception(f"Token exchange failed: {str(e)}")
    
    def refresh_access_token(self) -> Dict[str, Any]:
        """
        Refresh access token
        
        Returns:
            New token response dictionary
        """
        if not self.refresh_token:
            raise Exception("No refresh token available")
        
        try:
            response = requests.post(
                f"{self.api_url}/oauth/token",
                json={
                    'grant_type': 'refresh_token',
                    'refresh_token': self.refresh_token,
                    'client_id': self.client_id,
                    'client_secret': self.client_secret,
                }
            )
            response.raise_for_status()
            
            data = response.json()
            self.access_token = data['access_token']
            if 'refresh_token' in data:
                self.refresh_token = data['refresh_token']
            
            self._update_session_auth()
            
            return data
        except requests.exceptions.RequestException as e:
            raise Exception(f"Token refresh failed: {str(e)}")
    
    def set_tokens(self, access_token: str, refresh_token: Optional[str] = None):
        """
        Set access token manually
        
        Args:
            access_token: Access token
            refresh_token: Refresh token (optional)
        """
        self.access_token = access_token
        self.refresh_token = refresh_token
        self._update_session_auth()
    
    def _update_session_auth(self):
        """Update session with current access token"""
        if self.access_token:
            self.session.headers['Authorization'] = f"Bearer {self.access_token}"
    
    def _request(self, method: str, endpoint: str, **kwargs) -> Dict[str, Any]:
        """
        Make authenticated request with auto-retry on 401
        
        Args:
            method: HTTP method
            endpoint: API endpoint
            **kwargs: Additional request arguments
            
        Returns:
            Response JSON
        """
        url = f"{self.api_url}{endpoint}"
        
        try:
            response = self.session.request(method, url, **kwargs)
            
            # Auto-refresh on 401
            if response.status_code == 401 and self.refresh_token:
                self.refresh_access_token()
                response = self.session.request(method, url, **kwargs)
            
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            raise Exception(f"Request failed: {str(e)}")
    
    # =========================================================================
    # PMG NODE OPERATIONS (with RBAC)
    # =========================================================================
    
    def create_node(self, node: Dict[str, Any]) -> Dict[str, Any]:
        """
        Create a new PMG node
        
        Args:
            node: Node data with type, content, metadata, tags
            
        Returns:
            Created node data
        """
        payload = {
            'type': node['type'],
            'content': node.get('content', {}),
            'metadata': node.get('metadata', {}),
            'tags': node.get('tags', []),
            'source': 'plugin',
            'created_at': datetime.utcnow().isoformat(),
        }
        
        return self._request('POST', '/api/v1/nodes', json=payload)
    
    def get_node(self, node_id: str) -> Dict[str, Any]:
        """
        Read a PMG node by ID
        
        Args:
            node_id: Node ID
            
        Returns:
            Node data
        """
        return self._request('GET', f'/api/v1/nodes/{node_id}')
    
    def update_node(self, node_id: str, updates: Dict[str, Any]) -> Dict[str, Any]:
        """
        Update a PMG node
        
        Args:
            node_id: Node ID
            updates: Node updates
            
        Returns:
            Updated node data
        """
        return self._request('PATCH', f'/api/v1/nodes/{node_id}', json=updates)
    
    def delete_node(self, node_id: str) -> Dict[str, Any]:
        """
        Delete a PMG node
        
        Args:
            node_id: Node ID
            
        Returns:
            Deletion response
        """
        return self._request('DELETE', f'/api/v1/nodes/{node_id}')
    
    def query_nodes(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """
        Query PMG nodes with filters
        
        Args:
            query: Query parameters (type, tags, from_date, to_date, limit)
            
        Returns:
            List of nodes
        """
        params = {}
        if query:
            if 'type' in query:
                params['type'] = query['type']
            if 'tags' in query:
                params['tags'] = ','.join(query['tags'])
            if 'from_date' in query:
                params['from'] = query['from_date'].isoformat()
            if 'to_date' in query:
                params['to'] = query['to_date'].isoformat()
            if 'limit' in query:
                params['limit'] = query['limit']
        
        return self._request('GET', '/api/v1/nodes', params=params)
    
    def create_relationship(self, source_id: str, target_id: str, 
                          relationship_type: str) -> Dict[str, Any]:
        """
        Create relationship between nodes
        
        Args:
            source_id: Source node ID
            target_id: Target node ID
            relationship_type: Type of relationship
            
        Returns:
            Relationship data
        """
        payload = {
            'source_id': source_id,
            'target_id': target_id,
            'type': relationship_type,
        }
        
        return self._request('POST', '/api/v1/relationships', json=payload)
    
    # =========================================================================
    # AGENT REGISTRATION
    # =========================================================================
    
    def register_agent(self, agent: Dict[str, Any]) -> Dict[str, Any]:
        """
        Register a new agent
        
        Args:
            agent: Agent configuration with name, description, version, 
                   capabilities, endpoints
            
        Returns:
            Registered agent data
        """
        payload = {
            'name': agent['name'],
            'description': agent['description'],
            'version': agent['version'],
            'capabilities': agent.get('capabilities', []),
            'endpoints': {
                'webhook': agent['endpoints']['webhook'],
                'health': agent['endpoints']['health'],
            },
            'metadata': agent.get('metadata', {}),
            'registered_at': datetime.utcnow().isoformat(),
        }
        
        return self._request('POST', '/api/v1/agents/register', json=payload)
    
    def update_agent(self, agent_id: str, updates: Dict[str, Any]) -> Dict[str, Any]:
        """
        Update agent configuration
        
        Args:
            agent_id: Agent ID
            updates: Agent updates
            
        Returns:
            Updated agent data
        """
        return self._request('PATCH', f'/api/v1/agents/{agent_id}', json=updates)
    
    def unregister_agent(self, agent_id: str) -> Dict[str, Any]:
        """
        Unregister an agent
        
        Args:
            agent_id: Agent ID
            
        Returns:
            Unregistration response
        """
        return self._request('DELETE', f'/api/v1/agents/{agent_id}')
    
    def subscribe_to_events(self, agent_id: str, event_types: List[str]) -> Dict[str, Any]:
        """
        Subscribe to event types
        
        Args:
            agent_id: Agent ID
            event_types: List of event types to subscribe to
            
        Returns:
            Subscription response
        """
        payload = {'event_types': event_types}
        return self._request('POST', f'/api/v1/agents/{agent_id}/subscriptions', json=payload)
    
    # =========================================================================
    # EVENT HANDLING
    # =========================================================================
    
    def send_test_event(self, agent_id: str, event: Dict[str, Any]) -> Dict[str, Any]:
        """
        Send a test event
        
        Args:
            agent_id: Agent ID
            event: Event data with type and payload
            
        Returns:
            Event response
        """
        payload = {
            'type': event['type'],
            'payload': event['payload'],
            'timestamp': datetime.utcnow().isoformat(),
            'test': True,
        }
        
        return self._request('POST', f'/api/v1/agents/{agent_id}/test-event', json=payload)
    
    def send_event(self, event: Dict[str, Any]) -> Dict[str, Any]:
        """
        Send an event to the platform
        
        Args:
            event: Event data with type, payload, user_id
            
        Returns:
            Event response
        """
        payload = {
            'type': event['type'],
            'payload': event['payload'],
            'user_id': event['user_id'],
            'timestamp': datetime.utcnow().isoformat(),
            'source': 'plugin',
        }
        
        return self._request('POST', '/api/v1/events', json=payload)
    
    @staticmethod
    def verify_webhook_signature(payload: str, signature: str, secret: str) -> bool:
        """
        Verify webhook signature
        
        Args:
            payload: Request payload
            signature: Request signature header
            secret: Webhook secret
            
        Returns:
            True if signature is valid
        """
        expected_signature = hmac.new(
            secret.encode('utf-8'),
            payload.encode('utf-8'),
            hashlib.sha256
        ).hexdigest()
        
        return hmac.compare_digest(signature, expected_signature)
    
    # =========================================================================
    # UTILITY METHODS
    # =========================================================================
    
    def get_current_user(self) -> Dict[str, Any]:
        """
        Get current user info
        
        Returns:
            User data
        """
        return self._request('GET', '/api/v1/user/me')
    
    def check_agent_health(self, agent_id: str) -> Dict[str, Any]:
        """
        Check agent health
        
        Args:
            agent_id: Agent ID
            
        Returns:
            Health status
        """
        return self._request('GET', f'/api/v1/agents/{agent_id}/health')

File: sdk/nodejs/lifeos-sdk.js
/**
 * LifeOS Node.js SDK
 * Allows third-party developers to build plugins and agents
 * @version 1.0.0
 */

const axios = require('axios');
const crypto = require('crypto');

class LifeOSSDK {
  /**
   * Initialize the SDK
   * @param {Object} config - Configuration object
   * @param {string} config.apiUrl - LifeOS API URL
   * @param {string} config.clientId - OAuth client ID
   * @param {string} config.clientSecret - OAuth client secret
   * @param {string} config.redirectUri - OAuth redirect URI
   */
  constructor(config) {
    this.apiUrl = config.apiUrl || 'http://localhost:8000';
    this.clientId = config.clientId;
    this.clientSecret = config.clientSecret;
    this.redirectUri = config.redirectUri;
    this.accessToken = null;
    this.refreshToken = null;
    
    this.axiosInstance = axios.create({
      baseURL: this.apiUrl,
      timeout: 30000,
    });
    
    // Request interceptor for auth
    this.axiosInstance.interceptors.request.use(async (config) => {
      if (this.accessToken) {
        config.headers.Authorization = `Bearer ${this.accessToken}`;
      }
      return config;
    });
    
    // Response interceptor for token refresh
    this.axiosInstance.interceptors.response.use(
      (response) => response,
      async (error) => {
        if (error.response?.status === 401 && this.refreshToken) {
          await this.refreshAccessToken();
          return this.axiosInstance.request(error.config);
        }
        return Promise.reject(error);
      }
    );
  }

  // =========================================================================
  // AUTHENTICATION
  // =========================================================================

  /**
   * Get OAuth authorization URL
   * @param {string} state - Random state for CSRF protection
   * @returns {string} Authorization URL
   */
  getAuthorizationUrl(state) {
    const params = new URLSearchParams({
      response_type: 'code',
      client_id: this.clientId,
      redirect_uri: this.redirectUri,
      scope: 'read:nodes write:nodes register:agent send:events',
      state: state || crypto.randomBytes(16).toString('hex'),
    });
    
    return `${this.apiUrl}/oauth/authorize?${params.toString()}`;
  }

  /**
   * Exchange authorization code for access token
   * @param {string} code - Authorization code
   * @returns {Promise<Object>} Token response
   */
  async getAccessToken(code) {
    try {
      const response = await axios.post(`${this.apiUrl}/oauth/token`, {
        grant_type: 'authorization_code',
        code,
        client_id: this.clientId,
        client_secret: this.clientSecret,
        redirect_uri: this.redirectUri,
      });
      
      this.accessToken = response.data.access_token;
      this.refreshToken = response.data.refresh_token;
      
      return response.data;
    } catch (error) {
      throw new Error(`Token exchange failed: ${error.message}`);
    }
  }

  /**
   * Refresh access token
   * @returns {Promise<Object>} New token response
   */
  async refreshAccessToken() {
    try {
      const response = await axios.post(`${this.apiUrl}/oauth/token`, {
        grant_type: 'refresh_token',
        refresh_token: this.refreshToken,
        client_id: this.clientId,
        client_secret: this.clientSecret,
      });
      
      this.accessToken = response.data.access_token;
      if (response.data.refresh_token) {
        this.refreshToken = response.data.refresh_token;
      }
      
      return response.data;
    } catch (error) {
      throw new Error(`Token refresh failed: ${error.message}`);
    }
  }

  /**
   * Set access token manually
   * @param {string} accessToken - Access token
   * @param {string} refreshToken - Refresh token (optional)
   */
  setTokens(accessToken, refreshToken = null) {
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
  }

  // =========================================================================
  // PMG NODE OPERATIONS (with RBAC)
  // =========================================================================

  /**
   * Create a new PMG node
   * @param {Object} node - Node data
   * @param {string} node.type - Node type (document, note, task, etc)
   * @param {Object} node.content - Node content
   * @param {Object} node.metadata - Node metadata
   * @param {Array} node.tags - Node tags
   * @returns {Promise<Object>} Created node
   */
  async createNode(node) {
    try {
      const response = await this.axiosInstance.post('/api/v1/nodes', {
        type: node.type,
        content: node.content,
        metadata: node.metadata || {},
        tags: node.tags || [],
        source: 'plugin',
        created_at: new Date().toISOString(),
      });
      
      return response.data;
    } catch (error) {
      throw new Error(`Failed to create node: ${error.message}`);
    }
  }

  /**
   * Read a PMG node by ID
   * @param {string} nodeId - Node ID
   * @returns {Promise<Object>} Node data
   */
  async getNode(nodeId) {
    try {
      const response = await this.axiosInstance.get(`/api/v1/nodes/${nodeId}`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get node: ${error.message}`);
    }
  }

  /**
   * Update a PMG node
   * @param {string} nodeId - Node ID
   * @param {Object} updates - Node updates
   * @returns {Promise<Object>} Updated node
   */
  async updateNode(nodeId, updates) {
    try {
      const response = await this.axiosInstance.patch(`/api/v1/nodes/${nodeId}`, updates);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to update node: ${error.message}`);
    }
  }

  /**
   * Delete a PMG node
   * @param {string} nodeId - Node ID
   * @returns {Promise<Object>} Deletion response
   */
  async deleteNode(nodeId) {
    try {
      const response = await this.axiosInstance.delete(`/api/v1/nodes/${nodeId}`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to delete node: ${error.message}`);
    }
  }

  /**
   * Query PMG nodes with filters
   * @param {Object} query - Query parameters
   * @param {string} query.type - Node type filter
   * @param {Array} query.tags - Tag filters
   * @param {Date} query.fromDate - Date range start
   * @param {Date} query.toDate - Date range end
   * @param {number} query.limit - Result limit
   * @returns {Promise<Array>} List of nodes
   */
  async queryNodes(query = {}) {
    try {
      const params = new URLSearchParams();
      if (query.type) params.append('type', query.type);
      if (query.tags) params.append('tags', query.tags.join(','));
      if (query.fromDate) params.append('from', query.fromDate.toISOString());
      if (query.toDate) params.append('to', query.toDate.toISOString());
      if (query.limit) params.append('limit', query.limit);
      
      const response = await this.axiosInstance.get(`/api/v1/nodes?${params.toString()}`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to query nodes: ${error.message}`);
    }
  }

  /**
   * Create relationship between nodes
   * @param {string} sourceId - Source node ID
   * @param {string} targetId - Target node ID
   * @param {string} relationshipType - Type of relationship
   * @returns {Promise<Object>} Relationship data
   */
  async createRelationship(sourceId, targetId, relationshipType) {
    try {
      const response = await this.axiosInstance.post('/api/v1/relationships', {
        source_id: sourceId,
        target_id: targetId,
        type: relationshipType,
      });
      
      return response.data;
    } catch (error) {
      throw new Error(`Failed to create relationship: ${error.message}`);
    }
  }

  // =========================================================================
  // AGENT REGISTRATION
  // =========================================================================

  /**
   * Register a new agent
   * @param {Object} agent - Agent configuration
   * @param {string} agent.name - Agent name
   * @param {string} agent.description - Agent description
   * @param {string} agent.version - Agent version
   * @param {Array} agent.capabilities - Agent capabilities
   * @param {Object} agent.endpoints - Callback endpoints
   * @param {string} agent.endpoints.webhook - Webhook URL for events
   * @param {string} agent.endpoints.health - Health check URL
   * @returns {Promise<Object>} Registered agent data
   */
  async registerAgent(agent) {
    try {
      const response = await this.axiosInstance.post('/api/v1/agents/register', {
        name: agent.name,
        description: agent.description,
        version: agent.version,
        capabilities: agent.capabilities || [],
        endpoints: {
          webhook: agent.endpoints.webhook,
          health: agent.endpoints.health,
        },
        metadata: agent.metadata || {},
        registered_at: new Date().toISOString(),
      });
      
      return response.data;
    } catch (error) {
      throw new Error(`Failed to register agent: ${error.message}`);
    }
  }

  /**
   * Update agent configuration
   * @param {string} agentId - Agent ID
   * @param {Object} updates - Agent updates
   * @returns {Promise<Object>} Updated agent data
   */
  async updateAgent(agentId, updates) {
    try {
      const response = await this.axiosInstance.patch(`/api/v1/agents/${agentId}`, updates);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to update agent: ${error.message}`);
    }
  }

  /**
   * Unregister an agent
   * @param {string} agentId - Agent ID
   * @returns {Promise<Object>} Unregistration response
   */
  async unregisterAgent(agentId) {
    try {
      const response = await this.axiosInstance.delete(`/api/v1/agents/${agentId}`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to unregister agent: ${error.message}`);
    }
  }

  /**
   * Subscribe to event types
   * @param {string} agentId - Agent ID
   * @param {Array<string>} eventTypes - Event types to subscribe to
   * @returns {Promise<Object>} Subscription response
   */
  async subscribeToEvents(agentId, eventTypes) {
    try {
      const response = await this.axiosInstance.post(`/api/v1/agents/${agentId}/subscriptions`, {
        event_types: eventTypes,
      });
      
      return response.data;
    } catch (error) {
      throw new Error(`Failed to subscribe to events: ${error.message}`);
    }
  }

  // =========================================================================
  // EVENT HANDLING
  // =========================================================================

  /**
   * Send a test event
   * @param {string} agentId - Agent ID
   * @param {Object} event - Event data
   * @param {string} event.type - Event type
   * @param {Object} event.payload - Event payload
   * @returns {Promise<Object>} Event response
   */
  async sendTestEvent(agentId, event) {
    try {
      const response = await this.axiosInstance.post(`/api/v1/agents/${agentId}/test-event`, {
        type: event.type,
        payload: event.payload,
        timestamp: new Date().toISOString(),
        test: true,
      });
      
      return response.data;
    } catch (error) {
      throw new Error(`Failed to send test event: ${error.message}`);
    }
  }

  /**
   * Send an event to the platform
   * @param {Object} event - Event data
   * @param {string} event.type - Event type
   * @param {Object} event.payload - Event payload
   * @param {string} event.userId - User ID
   * @returns {Promise<Object>} Event response
   */
  async sendEvent(event) {
    try {
      const response = await this.axiosInstance.post('/api/v1/events', {
        type: event.type,
        payload: event.payload,
        user_id: event.userId,
        timestamp: new Date().toISOString(),
        source: 'plugin',
      });
      
      return response.data;
    } catch (error) {
      throw new Error(`Failed to send event: ${error.message}`);
    }
  }

  /**
   * Verify webhook signature
   * @param {string} payload - Request payload
   * @param {string} signature - Request signature header
   * @param {string} secret - Webhook secret
   * @returns {boolean} Signature is valid
   */
  verifyWebhookSignature(payload, signature, secret) {
    const expectedSignature = crypto
      .createHmac('sha256', secret)
      .update(payload)
      .digest('hex');
    
    return crypto.timingSafeEqual(
      Buffer.from(signature),
      Buffer.from(expectedSignature)
    );
  }

  // =========================================================================
  // UTILITY METHODS
  // =========================================================================

  /**
   * Get current user info
   * @returns {Promise<Object>} User data
   */
  async getCurrentUser() {
    try {
      const response = await this.axiosInstance.get('/api/v1/user/me');
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get user info: ${error.message}`);
    }
  }

  /**
   * Check agent health
   * @param {string} agentId - Agent ID
   * @returns {Promise<Object>} Health status
   */
  async checkAgentHealth(agentId) {
    try {
      const response = await this.axiosInstance.get(`/api/v1/agents/${agentId}/health`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to check agent health: ${error.message}`);
    }
  }
}

module.exports = LifeOSSDK;
